### DISCUSSION TITLE
How do I create a property in the user model that only a superuser can change?

### DISCUSSION THREAD

**Discussion by augusto-herrmann** [fastapi-users/fastapi-users#537] [Q&A] [answered]
Following the [documentation on the user model](https://frankie567.github.io/fastapi-users/configuration/model/), let's say we want to create a custom property (called `group`) to the `User` model that only a superuser can change:

```python
from fastapi_users import models


class User(models.BaseUser):
    group: int


class UserCreate(models.BaseUserCreate):
    group: int


class UserUpdate(User, models.BaseUserUpdate):
    group: int


class UserDB(User, models.BaseUserDB):
    pass

```

Doing it this way, the user can change their own `group` property by using `PATCH /user/me`, but we want to prohibit that.

I could add a `Depends` to restrict all of the `/user` routes, but that would be too restrictive as it wouldn't allow even a `GET` to the normal user.

```python
app.include_router(
    fastapi_users.get_users_router(),
    prefix="/users",
    tags=["users"],
    Depends(fastapi_users.current_user(active=True, superuser=True))
)
```

How can I restrict it so that only a superuser is allowed to do it, either of those:

a. change a specific `User` property, e.g., `group`, or
b. use `PATCH` on the `User` model at all?

The question is mostly about how to do 'a', but 'b' would also be enough for my current needs.

**Comment by frankie567** [comment, marked as answer]
Acutally, there are undocumented methods on the `User` model that control which properties are serialized before saving to database:

https://github.com/frankie567/fastapi-users/blob/cdb62bfdf7041e6069389ca5d021ca58235a7978/fastapi_users/models.py#L7-L21

Basically, `PATCH /me` calls `create_update_dict`, while `PATCH /{user_id}` (for superusers) calls `create_update_dict_superuser`. What I do in my projects to implement your use case is to overload those methods to remove the properties I don't want to keep:

```py
classe User(models.BaseUser):
  group: int

  def create_update_dict(self):
      d = super().create_update_dict()
      d.pop("group", None)
      return d
```

**Reply by frankie567** [reply]
Maybe this could be a feature to integrate in the library? 🤔

## First idea 
Instead of overloading those methods, we could have properties like `_exclude` and `_exclude_superuser` on the model that the end developer could set to clearly state which properties they want to ignore.

Django REST Framework has something similar: https://www.django-rest-framework.org/api-guide/serializers/#specifying-which-fields-to-include

## Second idea

Maybe have a `BaseUserUpdateSuperuser` model that will have the properties updatable by a superuser. That would be more "FastAPI-idiomatic".

I have to think about this.

**Reply by augusto-herrmann** [reply]
Thanks for the quick reply, @frankie567 !

I did overload the `create_update_dict` method as you suggested, and indeed the database is not changed with the user tries `PATCH /user/me`. However, [and I have a test for this](https://github.com/economiagovbr/api-pgd/blob/main/test_api.py#L301), the API still returns the HTTP code 200, not 401 Unauthorized. I suppose I could raise an exception inside `create_update_dict` when the prohibited property is present in the dict,

```python
  def create_update_dict(self):
      d = super().create_update_dict()
      p = d.pop("group", None)
      if p:
          raise HTTPException(401, detail=f"Not authorized to change property {p}")
      return d
```

but I think that would be quite hacky and not the appropriate solution. How could I return 401 appropriately in this situation?

As for the feature ideas, they are both interesting, but I don't really have an opinion to favor one over the other.

**Reply by frankie567** [reply]
Hum, why would you want to raise a 401 error? The default behaviour of Pydantic is to silently ignore the properties it doesn't know, so it's quite consistent here.

If you really want to raise an error here, you could do what you show here, even if I agree it's not very elegant ; or re-implement the route yourself with proper logic.

**Reply by augusto-herrmann** [reply]
I want to raise a 401 error because the user is not allowed to change the (`group`, in this example) property. If they try to do

```
PATCH /users/me

{
    "group": 9999
}
```

then it seems to me that the expected behaviour should definitely be a 401, because the attempt by the user to change the property is not authorized. Not returning status code 200 with the complete user object and the `group` property unchanged in the body of the request.

For now, I'm going with the inelegant solution. If you think it could be valuable, I might re-implement the route and create a PR later if I have the time.

**Reply by augusto-herrmann** [reply]
Thanks for the help, @frankie567 ! I'm marking this as solved.

**Reply by frankie567** [reply]
Personally, I don't mind having a 200 status for this use-case. The route `/users/me` should just ignore this parameter, as it would do for any parameter it doesn't know.

The only drawback is that the automatic Swagger documentation will show `group` as an input field, which makes me think that I'll probably work on having the possibility to have two different Pydantic models for regular users and superusers.

**Reply by augusto-herrmann** [reply]
Actually, for my use case the regular user can **read**, but not **write** the specific property `group`. I don't know if Pydantic supports read-only attributes, but it might be something worth looking into.
